### PR TITLE
support tekton-results

### PR DESCRIPTION
**Pull request by christianvogt**
## Fixes 
https://issues.redhat.com/browse/HAC-2568

## Description
Adds support to fetch pipeline runs and tasks runs from tekton results in order to combine the results with those from etcd. This is necessary because runs are actively purged from etcd.

Note that logs are not yet retrieved from tekton results.

Aside from logs, there should be no visual differences for runs coming from tekton results and those from etcd.

The main files that support this feature are:
- `utils/tekton-results.ts`
  - Utilities for querying tekton-results.
- `hooks/useTektonResults.ts`
  - Hook interface to `tekton-results.ts`
- `hooks/usePipelineRuns.ts`
  - This does the heavy lifting of fetching from etcd and tekton results, deduping results, and providing the final results.

## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 

image

## How to test or reproduce?

Create many pipeline runs (> 10) and wait for them to be purged from etcd.

## Browser conformance: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by christianvogt** [comment]
2nd commit adds utilities for working with logs from tekton results. Although these utilities aren't used anywhere other than unit tests.

File: src/hooks/usePACState.ts [hunk 1]
```diff
@@ -19,34 +19,35 @@ const usePACState = (component: ComponentKind) => {
   const isSample = component.metadata?.annotations?.[SAMPLE_ANNOTATION] === 'true';
   const pacProvision = getPACProvision(component);
 
+  const { name: PR_BOT_NAME } = useApplicationPipelineGitHubApp();
```
Comment (jeff-phillips-18) [inline src/hooks/usePACState.ts:22]: All caps make me think it's a constant defined somewhere.

File: src/hooks/usePipelineRuns.ts [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+  // if a pipeline run was removed from etcd, we want to still include it in the return value without re-querying tekton-results
+  const etcdRuns = React.useMemo(
+    () => (isList ? result[0] : result[1] ? [result[0]] : []) as Kind[],
+    [isList, result],
+  );
+
+  const etcdRunsRef = React.useRef<Kind[]>([]);
+  const runs = React.useMemo(() => {
+    if (!etcdRuns) {
```
Comment (jeff-phillips-18) [inline src/hooks/usePipelineRuns.ts:52]: Better safe than sorry, but shouldn't occur.

File: src/hooks/useTektonResults.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        } catch (e) {
+          if (nextPageToken) {
+            setResult((cur) => [cur[0], cur[1], e, undefined]);
+          } else {
+            setResult([[], false, e, undefined]);
+          }
+        }
+      })();
+    }
+  }, [namespace, options, nextPageToken, cacheKey, getRuns]);
```
Comment (jeff-phillips-18) [inline src/hooks/useTektonResults.ts:74]: Need a destructor to set a `canceled` flag so as to not update state if effect is destroyed.

File: src/hooks/usePipelineRuns.ts [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    return [
+      rResources,
+      namespace
+        ? queryTr
+          ? isList
+            ? // return loaded only if both sources have loaded
+              trLoaded && loaded
+            : // when searching by name, loading fails if we have no result
+              !!rResources?.[0]
+          : isList
```
Comment (jeff-phillips-18) [inline src/hooks/usePipelineRuns.ts:108]: Could `isList` be true here? Too much ternary makes my brain melt.

File: src/hooks/usePipelineRuns.ts [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+          ? loaded
+          : // when searching by name, loading fails if we have no result
+            !!rResources?.[0]
+        : false,
```
Comment (jeff-phillips-18) [inline src/hooks/usePipelineRuns.ts:112]: Should this be `loaded`?

File: src/hooks/usePipelineRuns.ts [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+          selector: optionsMemo?.selector,
+          name: optionsMemo?.name,
+        }
+      : null;
+  }, [groupVersionKind, namespace, optionsMemo, isList]);
+
+  const [resources, loaded, error] = useK8sWatchResource(watchOptions);
+  // if a pipeline run was removed from etcd, we want to still include it in the return value without re-querying tekton-results
+  const etcdRuns = React.useMemo(
+    () => (isList ? resources : loaded ? [resources] : []) as Kind[],
```
Comment (jeff-phillips-18) [inline src/hooks/usePipelineRuns.ts:47]:
```suggestion
    () => (isList ? resources : loaded && !error ? [resources] : []) as Kind[],
```

File: src/hooks/usePipelineRuns.ts (hunk 5 above)
Comment (christianvogt) [reply src/hooks/usePipelineRuns.ts:112]: we return `false` when there is no `namespace`

File: src/hooks/usePipelineRuns.ts (hunk 6 above)
Comment (christianvogt) [reply src/hooks/usePipelineRuns.ts:47]: it's very odd that `useK8sWatchResource` can return `[{}, true, error]`. Why an empty object when we have an error.
I've added a test case for this and updated to the suggestion. All other test cases still pass.

**Comment by christianvogt** [comment]
@jeff-phillips-18 I've made all the changes

File: src/utils/tekton-results.ts [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+export const getTaskRuns = (
+  namespace: string,
+  options?: TektonResultsOptions,
+  nextPageToken?: string,
+  // supply a cacheKey only if the TaskRun is complete and response will never change in the future
+  cacheKey?: string,
+) => getFilteredTaskRuns(namespace, '', options, nextPageToken, cacheKey);
+
+const getLog = (taskRunPath: string) =>
+  commonFetchJSON<Log>(`${URL_PREFIX}/parents/${taskRunPath.replace('/records/', '/logs/')}`);
```
Comment (jeff-phillips-18) [inline src/utils/tekton-results.ts:283]:
```suggestion
  commonFetchJSON<Log>(`${URL_PREFIX}/${taskRunPath.replace('/records/', '/logs/')}`);
```

**Comment by christianvogt** [comment]
@jeff-phillips-18 rebased again

**Comment by christianvogt** [comment]
issue with tekton-results throwing error 500

**Pull request merged**